### ISSUE TITLE
RegisterGradient not behaving as expected.

### ISSUE THREAD

**Issue by GeneKitamura** [tensorflow/tensorflow#29449]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- TensorFlow installed from (source or binary): Docker
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5
- CUDA/cuDNN version: CUDA 9.0
- GPU model and memory: NVIDIA GTX1070 8GB, 32 GB RAM

So I'm trying to modify the tensorflow gradients as it flows through the graph, but I'm getting peculiar results. So as a simple example, if I change the "Relu" gradient to my "Custom" gradient, which should just pass the gradient through, I get an array of all 1's, rather than expected correct gradient. 

```
import tensorflow as tf
import numpy as np

graph = tf.get_default_graph()
sess = tf.Session(graph=graph)

@tf.RegisterGradient("Custom")
def _no_modifications(unused_op, grad):
    return grad

c_input = tf.Variable((np.arange(10).reshape(5, 2) - 5), dtype=tf.float32)

with graph.gradient_override_map({"Relu": "Custom"}):
    c_output = tf.nn.relu(c_input)

grad = tf.gradients(c_output, c_input)

sess.run(tf.global_variables_initializer())

input_val, output_val, grad_val = sess.run([c_input, c_output, grad])

input_val
array([[-5., -4.],
       [-3., -2.],
       [-1.,  0.],
       [ 1.,  2.],
       [ 3.,  4.]], dtype=float32)

output_val
array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [1., 2.],
       [3., 4.]], dtype=float32)

grad_val #This is incorrect
[array([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]], dtype=float32)]


#The correct gradient should be:
[array([[0., 0.],
        [0., 0.],
        [0., 0.],
        [1., 1.],
        [1., 1.]], dtype=float32)]

```
